### PR TITLE
Check if all dimensions in output are non-zero

### LINKED ISSUE #48589: Conv2d didn't raise exception for invalid input argument.
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): bianry
- TensorFlow version (use command below): 2 .4.1
- Python version: 3.7

**Standalone code to reproduce the issue**
When `Conv2D` with `kernel_size`=`2` and padding=`valid` receives an invalid input, it does not raise any exception. Instead it outputs a tensor with zero-dimension. This can lead to future crash for other APIs with 0-dim tensor as input.

```
import tensorflow as tf
import numpy as np

filters, kernel_size, strides, padding = 3, [2, 2], 2, 'valid'
data = np.random.rand(1, 1, 1, 1)
layer = tf.keras.layers.Conv2D(filters, kernel_size, strides=strides, padding=padding)
print(layer(data).shape)
```

Outputs
```
(1, 0, 0, 3)
```

**Describe the current behavior**
No exception is raised for invalid input argument.

**Describe the expected behavior**
Expect `ValueError` to be raised.

### PR DESCRIPTION
**Pull request by AdityaKane2001**
See #48589

### CODE REVIEW

**Review by irvifa — COMMENTED** [review]
Is it possible to add the test on `local_test.py` as well as `convolutional_test.py`?

**Comment by AdityaKane2001** [comment]
@irvifa 
Yes, I'm working on that.

**Comment by AdityaKane2001** [comment]
@irvifa 
Made requested additions.

**Comment by AdityaKane2001** [comment]
@gbaned 
Requesting to review this one. Thanks!

File: tensorflow/python/keras/layers/local.py [hunk 1]
```diff
@@ -729,14 +729,14 @@ def local_conv_matmul(inputs, kernel, kernel_mask, output_shape):
   Returns:
       Output (N+2)-D tensor with shape `output_shape`.
   """
-  inputs_flat = backend.reshape(inputs, (backend.shape(inputs)[0], -1))
+  inputs_flat = array_ops.reshape(inputs, (array_ops.shape(inputs)[0], -1))
```
Comment (irvifa) [inline tensorflow/python/keras/layers/local.py:732]: I'm wondering if we can somehow mentioned on why there's changes on the implementation on the PR's description, previously we're using `backend` instead of `array_ops`.

Comment (AdityaKane2001) [reply tensorflow/python/keras/layers/local.py:732]: @irvifa 
Can you please elaborate on that?

Comment (AdityaKane2001) [reply tensorflow/python/keras/layers/local.py:732]: Actually in the issue the user pointed out that some of the calls were deprecated. Also, backend also calls `array_ops` methods, so I called them directly.

File: tensorflow/python/keras/layers/convolutional.py [hunk 2]
```diff
@@ -230,6 +230,15 @@ def build(self, input_shape):
     tf_op_name = self.__class__.__name__
     if tf_op_name == 'Conv1D':
       tf_op_name = 'conv1d'  # Backwards compat.
+    
+    # Check if output shapes are valid
+    # They must not have 0 entries along any dimension
+    output_shapes = self.compute_output_shape(input_shape).as_list()
```
Comment (chenmoneygithub) [inline tensorflow/python/keras/layers/convolutional.py:236]: Please change to singular form: output_shapes => output_shape

File: tensorflow/python/keras/layers/convolutional.py [hunk 3]
```diff
@@ -230,6 +230,15 @@ def build(self, input_shape):
...
+    output_shapes = list(filter((None).__ne__, output_shapes))
+    if not all(output_shapes):
```
Comment (chenmoneygithub) [inline tensorflow/python/keras/layers/convolutional.py:238]: all() operator here is a bit unclear, please change to 0 in output_shapes, or do specific checks on the dimension.

File: tensorflow/python/keras/layers/convolutional_test.py [hunk 4]
```diff
@@ -298,6 +303,11 @@ def test_conv2d_zero_kernel_size(self):
     with self.assertRaises(ValueError):
       keras.layers.Conv2D(**kwargs)
 
+  def test_conv2d_valid_output_shapes(self):
```
Comment (chenmoneygithub) [inline tensorflow/python/keras/layers/convolutional_test.py:306]: Same as above

**Review by chenmoneygithub — COMMENTED** [review]
Thanks for the PR! Looks good overall, dropped some comments on the style.

**Comment by AdityaKane2001** [comment]
@chenmoneygithub 
Thanks for the review! I have made the requested changes.

**Review by chenmoneygithub — APPROVED** [review]
Thanks for the PR!

**Comment by AdityaKane2001** [comment]
@chenmoneygithub 
Thanks for the approval!

**Comment by AdityaKane2001** [comment]
@chenmoneygithub 
The build tests failed. I will fix the errors and then request a review.
Thanks.

**Comment by AdityaKane2001** [comment]
@chenmoneygithub 
Requesting review again. I have fixed the syntactic bugs that were present. Using `conv_utils.output_length` function, which does not return `TensorShape` instance for easier debugging.

File: tensorflow/python/keras/layers/convolutional.py [hunk 5]
```diff
@@ -231,6 +231,33 @@ def build(self, input_shape):
     if tf_op_name == 'Conv1D':
       tf_op_name = 'conv1d'  # Backwards compat.
 
+    # Check if output shapes are valid
+    # They must not have 0 entries along any dimension
+    if self._channels_first:
```
Comment (chenmoneygithub) [inline tensorflow/python/keras/layers/convolutional.py:236]: Could you explain why we need to skip checks on the channel axis? Also please consider putting a comment before the if check.

File: tensorflow/python/keras/layers/convolutional.py [hunk 6]
```diff
@@ -231,6 +231,33 @@ def build(self, input_shape):
...
+      for idx,dimension in enumerate(input_shape.as_list()[2:]):
```
Comment (chenmoneygithub) [inline tensorflow/python/keras/layers/convolutional.py:237]: The code chunk under if and else are basically same, usually we discourage duplicated code. Could you create a small helper function to wrap the code? Thx!

Comment (AdityaKane2001) [reply tensorflow/python/keras/layers/convolutional.py:237]: Sure, I'll add one

File: tensorflow/python/keras/layers/convolutional.py (hunk 5 above)
Comment (AdityaKane2001) [reply tensorflow/python/keras/layers/convolutional.py:236]: That missed that. I'll fix the code and push again. Thanks for the inputs.

Comment (AdityaKane2001) [reply tensorflow/python/keras/layers/convolutional.py:236]: Also the rest of the arguments for `conv_utils.conv_output_length` don't consider channels and batch size.

**Comment by AdityaKane2001** [comment]
@chenmoneygithub
Made requested changes

**Comment by AdityaKane2001** [comment]
@chenmoneygithub 
Made requested changes

**Comment by AdityaKane2001** [comment]
@chenmoneygithub 
Requesting review.

File: tensorflow/python/keras/layers/local_test.py [hunk 7]
```diff
@@ -158,6 +158,11 @@ def test_locallyconnected_1d_regularization(self, data_format, padding,
         self.assertEqual(layer.kernel.constraint, k_constraint)
         self.assertEqual(layer.bias.constraint, b_constraint)
 
+  def test_locallyconnected1d_invalid_output_shapes(self):
+    kwargs = {'filters': 2 , 'kernel_size': 10}
```
Comment (chenmoneygithub) [inline tensorflow/python/keras/layers/local_test.py:162]: Nit: delete the space after 'filters': 2. 'filters': 2 , => 'filters': 2,

File: tensorflow/python/keras/layers/local_test.py [hunk 8]
```diff
@@ -265,6 +270,12 @@ def test_locallyconnected_2d_regularization(self, data_format, padding,
         self.assertEqual(layer.kernel.constraint, k_constraint)
         self.assertEqual(layer.bias.constraint, b_constraint)
 
+  def test_locallyconnected2d_invalid_output_shapes(self):
+    kwargs = {'filters': 2 , 'kernel_size': 10}
```
Comment (chenmoneygithub) [inline tensorflow/python/keras/layers/local_test.py:274]: Same here

File: tensorflow/python/keras/layers/convolutional_test.py [hunk 9]
```diff
@@ -551,6 +571,11 @@ def test_conv3d_transpose(self, kwargs, expected_output_shape=None):
     if 'data_format' not in kwargs or test.is_gpu_available(cuda_only=True):
       self._run_test(kwargs, expected_output_shape)
 
+  def test_conv3dtanspose_invalid_output_shapes(self):
+    kwargs = {'filters': 2 , 'kernel_size': 10}
```
Comment (chenmoneygithub) [inline tensorflow/python/keras/layers/convolutional_test.py:575]: Same here

File: tensorflow/python/keras/layers/convolutional_test.py [hunk 10]
```diff
@@ -518,6 +533,11 @@ def test_conv1d_transpose(self, kwargs, expected_output_shape=None):
         test.is_gpu_available(cuda_only=True)):
       self._run_test(kwargs, expected_output_shape)
 
+  def test_conv1dtranspose_invalid_output_shapes(self):
+    kwargs = {'filters': 2 , 'kernel_size': 10}
```
Comment (chenmoneygithub) [inline tensorflow/python/keras/layers/convolutional_test.py:537]: Same here

File: tensorflow/python/keras/layers/convolutional.py [hunk 11]
```diff
@@ -303,6 +314,19 @@ def compute_output_shape(self, input_shape):
...
     return False
 
+  def _check_invalid_dimension(self, dimension, idx):
+    """Checks if output has all positive dimensions"""
+    output_dimension = conv_utils.conv_output_length(dimension,
+                                                     self.kernel_size[idx],
+                                                     self.padding,
+                                                     self.strides[idx],
+                                                     dilation = self.dilation_rate[idx])
+    if output_dimension is not None and output_dimension<=0:
```
Comment (chenmoneygithub) [inline tensorflow/python/keras/layers/convolutional.py:324]: Nit: space between var and operator: output_dimension<=0 => output_dimension <= 0

**Review by chenmoneygithub — COMMENTED** [review]
Looks good overall, left some comments on the code style.

**Comment by AdityaKane2001** [comment]
@chenmoneygithub 
Made requested changes. Requesting review.

**Review by chenmoneygithub — APPROVED** [review]

**Comment by rthadur** [comment]
@AdityaKane2001 can you please check sanity build failures ?

**Comment by AdityaKane2001** [comment]
@rthadur 
Yes

**Comment by AdityaKane2001** [comment]
@rthadur
```python
  def _check_invalid_dimension(self, dimension, idx):
    """Checks if output has all positive dimensions"""
    output_dimension = conv_utils.conv_output_length(dimension,
                                                     self.kernel_size[idx],
                                                     self.padding,
                                                     self.strides[idx],
                                                     dilation = self.dilation_rate[idx]) #line 323
```
In `convolutional.py` the line 323 gives sanity error. If I delete the spaces before `dilation` then it may give another sanity error. Should I disable pylint long line check for this line only? Or is there some other solution?

**Comment by rthadur** [comment]
@AdityaKane2001 please try below , thank you 

```
def _check_invalid_dimension(self, dimension, idx):
    """Checks if output has all positive dimensions."""
    output_dimension = conv_utils.conv_output_length(
        dimension,
        self.kernel_size[idx],
        self.padding,
        self.strides[idx],
        dilation=self.dilation_rate[idx])
```

**Review by rthadur — APPROVED** [review]

**Comment by AdityaKane2001** [comment]
@rthadur 
The last commit should fix the build errors. Could you please run the tests again? Thanks

**Review by rthadur — APPROVED** [review]

**Comment by AdityaKane2001** [comment]
@rthadur
Thanks for the review

**Comment by AdityaKane2001** [comment]
@rthadur 
I believe the windows bazel build error is persisting for all of the recent PRs.

**Comment by AdityaKane2001** [comment]
@rthadur 
I just synced the repo to my fork, and I saw that PRs were rolled back by tensorflower-gardener. Can you please explain what's happening? Commit [here](https://github.com/tensorflow/tensorflow/commit/0174bcb337e2c8e7d6782b60b094ea70460ad071#diff-83c641ba1ee058bf5b2afe821cedf89f177973095960a14739ef12a9575416b3).
